### PR TITLE
feat: [PLG-2382]: Added creditCard resource and service

### PR DESCRIPTION
**Pull request by thomas-grasman**
## Describe your changes
Added new credit card entity and service for validating and saving credit cards.
No sensitive card information is stored, only the fingerprint (hash value) of the credit card.

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/48199)

### CODE REVIEW

**Comment by thomas-grasman** [comment]
trigger uts

File: 930-ng-license-manager/src/test/java/io/harness/subscription/services/CreditCardServiceImplTest.java [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+    creditCardService = new CreditCardServiceImpl(creditCardRepository);
+  }
+
+  @Test
+  @Owner(developers = TOMMY)
+  @Category(UnitTests.class)
+  public void testSaveCreditCard() {
+    when(creditCardRepository.findByFingerprint(DEFAULT_FINGERPRINT)).thenReturn(DEFAULT_CREDIT_CARD);
+
+    creditCardService.saveCreditCard(DEFAULT_ACCOUNT_ID, DEFAULT_FINGERPRINT);
```
Comment (Xin-Harness) [inline 930-ng-license-manager/src/test/java/io/harness/subscription/services/CreditCardServiceImplTest.java:49]: Missing assert here

**Review by kapil-harness — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/48199)* status: 9 file(s) to review. 2 change requests. 0 of 2 approvals obtained. 

___
*[`930-ng-license-manager/src/main/java/io/harness/subscription/resource/CreditCardResource.java` line 75 at r3](https://reviewable.io/reviews/harness/harness-core/48199#-NWBdQcL8NQkSZGIMEh7:-NWBdQcMBs_QWC8b9p5R:bqtb9nq) ([raw file](https://github.com/harness/harness-core/blob/e7d529998df23d751d4da77c912f11971ea6d158/930-ng-license-manager/src/main/java/io/harness/subscription/resource/CreditCardResource.java#L75)):*
> ```Java
>           required = true, description = "Unique fingerprint of the credit card") @NotNull @Valid String fingerPrint) {
>     creditCardService.saveCreditCard(accountIdentifier, fingerPrint);
>     return ResponseDTO.newResponse();
> ```

lets not return empty response, lets construct a response and return that e.g. io.harness.cvng.notification.beans.NotificationRuleResponse and API is io.harness.cvng.notification.resources.NotificationRuleResource#saveNotificationRuleData
___
*[`930-ng-license-manager/src/main/java/io/harness/subscription/services/impl/CreditCardServiceImpl.java` line 29 at r3](https://reviewable.io/reviews/harness/harness-core/48199#-NWBdrTrE4DcXPQHC2Vk:-NWBdrTrE4DcXPQHC2Vl:bnfbkx) ([raw file](https://github.com/harness/harness-core/blob/e7d529998df23d751d4da77c912f11971ea6d158/930-ng-license-manager/src/main/java/io/harness/subscription/services/impl/CreditCardServiceImpl.java#L29)):*
> ```Java
>     CreditCard creditCard = creditCardRepository.findByFingerprint(fingerprint);
> 
>     if (creditCard != null && !creditCard.getAccountIdentifier().equals(accountIdentifier)) {
> ```

not a good way of writing a CRUD API
___
*[`930-ng-license-manager/src/main/java/io/harness/subscription/services/impl/CreditCardServiceImpl.java` line 30 at r3](https://reviewable.io/reviews/harness/harness-core/48199#-NWBfeuVEE4Hd5ubcUfk:-NWBfeuVEE4Hd5ubcUfl:bk4p2c3) ([raw file](https://github.com/harness/harness-core/blob/e7d529998df23d751d4da77c912f11971ea6d158/930-ng-license-manager/src/main/java/io/harness/subscription/services/impl/CreditCardServiceImpl.java#L30)):*
> ```Java
> 
>     if (creditCard != null && !creditCard.getAccountIdentifier().equals(accountIdentifier)) {
>       throw new BadRequestException("Could not save credit card.");
> ```

lets throw DuplicateFieldException when creditCard != null
for the 2nd condition you can throw some other exception i.e. !creditCard.getAccountIdentifier().equals(accountIdentifier)
___
*[`930-ng-license-manager/src/main/java/io/harness/subscription/services/impl/CreditCardServiceImpl.java` line 33 at r3](https://reviewable.io/reviews/harness/harness-core/48199#-NWBg9L60ARlSeOjHi6o:-NWBg9L70_YLBn_B3mZa:b-icof1) ([raw file](https://github.com/harness/harness-core/blob/e7d529998df23d751d4da77c912f11971ea6d158/930-ng-license-manager/src/main/java/io/harness/subscription/services/impl/CreditCardServiceImpl.java#L33)):*
> ```Java
>     }
> 
>     if (creditCard == null) {
> ```

this is not required once we have checked for DuplicateFieldException above
___
*[`930-ng-license-manager/src/main/java/io/harness/subscription/services/impl/CreditCardServiceImpl.java` line 34 at r3](https://reviewable.io/reviews/harness/harness-core/48199#-NWBgKa1DRVSgGYR3UnZ:-NWBgKa1DRVSgGYR3Un_:b-1aq47c) ([raw file](https://github.com/harness/harness-core/blob/e7d529998df23d751d4da77c912f11971ea6d158/930-ng-license-manager/src/main/java/io/harness/subscription/services/impl/CreditCardServiceImpl.java#L34)):*
> ```Java
> 
>     if (creditCard == null) {
>       creditCardRepository.save(
> ```

lets put accountID and fingerprint inside a DTO and get this DTO as an input from API arguments.
post that pls create DTO to entity object mappers. Code reference:



_Code snippet:_
```Java
io.harness.cvng.notification.services.impl.NotificationRuleServiceImpl#create
```
___
*[`930-ng-license-manager/src/test/java/io/harness/subscription/services/CreditCardServiceImplTest.java` line 58 at r3](https://reviewable.io/reviews/harness/harness-core/48199#-NWBjpcZ73qfMHZQoiH9:-NWBjpcZ73qfMHZQoiHA:b-gs0zn9) ([raw file](https://github.com/harness/harness-core/blob/e7d529998df23d751d4da77c912f11971ea6d158/930-ng-license-manager/src/test/java/io/harness/subscription/services/CreditCardServiceImplTest.java#L58)):*
> ```Java
>     when(creditCardRepository.findByFingerprint(DEFAULT_FINGERPRINT)).thenReturn(DEFAULT_CREDIT_CARD);
> 
>     creditCardService.saveCreditCard(ALTERNATE_ACCOUNT_ID, DEFAULT_FINGERPRINT);
> ```

Missing assert here

**Comment by thomas-grasman** [comment]
trigger uts

**Comment by thomas-grasman** [comment]
trigger uts

**Review by thomas-grasman — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/48199)* status: 14 file(s) to review. 2 change requests. 8. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`930-ng-license-manager/src/main/java/io/harness/subscription/resource/CreditCardResource.java` line 72 at r3](https://reviewable.io/reviews/harness/harness-core/48199#gh-1202524224:-NWEJPJb2Brl-kteqmuC:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/e7d529998df23d751d4da77c912f11971ea6d158/930-ng-license-manager/src/main/java/io/harness/subscription/resource/CreditCardResource.java#L72)):*
<details><summary><i>Previously, Xin-Harness (Xin Shao) wrote…</i></summary><blockquote>

Recommand to use @RequestBody combine with import io.swagger.v3.oas.annotations.parameters.RequestBody instead of inline package import
</blockquote></details>

Done.
___
*[`930-ng-license-manager/src/main/java/io/harness/subscription/resource/CreditCardResource.java` line 75 at r3](https://reviewable.io/reviews/harness/harness-core/48199#-NWBdQcL8NQkSZGIMEh7:-NWEJPrrA-NvcQpQRHyf:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/e7d529998df23d751d4da77c912f11971ea6d158/930-ng-license-manager/src/main/java/io/harness/subscription/resource/CreditCardResource.java#L75)):*
<details><summary><i>Previously, kapil-harness (Kapil Choudhary) wrote…</i></summary><blockquote>

lets not return empty response, lets construct a response and return that e.g. io.harness.cvng.notification.beans.NotificationRuleResponse and API is io.harness.cvng.notification.resources.NotificationRuleResource#saveNotificationRuleData
</blockquote></details>

Done.
___
*[`930-ng-license-manager/src/main/java/io/harness/subscription/services/impl/CreditCardServiceImpl.java` line 29 at r3](https://reviewable.io/reviews/harness/harness-core/48199#-NWBdrTrE4DcXPQHC2Vk:-NWEJR2X5IrnJo7oMjMY:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/e7d529998df23d751d4da77c912f11971ea6d158/930-ng-license-manager/src/main/java/io/harness/subscription/services/impl/CreditCardServiceImpl.java#L29)):*
<details><summary><i>Previously, kapil-harness (Kapil Choudhary) wrote…</i></summary><blockquote>

not a good way of writing a CRUD API
</blockquote></details>

Done.
___
*[`930-ng-license-manager/src/main/java/io/harness/subscription/services/impl/CreditCardServiceImpl.java` line 30 at r3](https://reviewable.io/reviews/harness/harness-core/48199#-NWBfeuVEE4Hd5ubcUfk:-NWEJRKc1vJyKEaODsA_:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/e7d529998df23d751d4da77c912f11971ea6d158/930-ng-license-manager/src/main/java/io/harness/subscription/services/impl/CreditCardServiceImpl.java#L30)):*
<details><summary><i>Previously, kapil-harness (Kapil Choudhary) wrote…</i></summary><blockquote>

lets throw DuplicateFieldException when creditCard != null
for the 2nd condition you can throw some other exception i.e. !creditCard.getAccountIdentifier().equals(accountIdentifier)
</blockquote></details>

Done.
___
*[`930-ng-license-manager/src/main/java/io/harness/subscription/services/impl/CreditCardServiceImpl.java` line 33 at r3](https://reviewable.io/reviews/harness/harness-core/48199#-NWBg9L60ARlSeOjHi6o:-NWEJSPN1FHBHxN58fu6:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/e7d529998df23d751d4da77c912f11971ea6d158/930-ng-license-manager/src/main/java/io/harness/subscription/services/impl/CreditCardServiceImpl.java#L33)):*
<details><summary><i>Previously, kapil-harness (Kapil Choudhary) wrote…</i></summary><blockquote>

this is not required once we have checked for DuplicateFieldException above
</blockquote></details>

Done.
___
*[`930-ng-license-manager/src/main/java/io/harness/subscription/services/impl/CreditCardServiceImpl.java` line 34 at r3](https://reviewable.io/reviews/harness/harness-core/48199#-NWBgKa1DRVSgGYR3UnZ:-NWEJTF91JpS5BkIYHy4:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/e7d529998df23d751d4da77c912f11971ea6d158/930-ng-license-manager/src/main/java/io/harness/subscription/services/impl/CreditCardServiceImpl.java#L34)):*
<details><summary><i>Previously, kapil-harness (Kapil Choudhary) wrote…</i></summary><blockquote>

lets put accountID and fingerprint inside a DTO and get this DTO as an input from API arguments.
post that pls create DTO to entity object mappers. Code reference:

</blockquote></details>

Done.
___
*[`930-ng-license-manager/src/test/java/io/harness/subscription/services/CreditCardServiceImplTest.java` line 49 at r3](https://reviewable.io/reviews/harness/harness-core/48199#gh-1202527923:-NWEJU9N9cT4cFYsFnZK:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/e7d529998df23d751d4da77c912f11971ea6d158/930-ng-license-manager/src/test/java/io/harness/subscription/services/CreditCardServiceImplTest.java#L49)):*
<details><summary><i>Previously, Xin-Harness (Xin Shao) wrote…</i></summary><blockquote>

Missing assert here
</blockquote></details>

Done.
___
*[`930-ng-license-manager/src/test/java/io/harness/subscription/services/CreditCardServiceImplTest.java` line 58 at r3](https://reviewable.io/reviews/harness/harness-core/48199#-NWBjpcZ73qfMHZQoiH9:-NWEJW430n20OCWWxHlh:bnqsjc9) ([raw file](https://github.com/harness/harness-core/blob/e7d529998df23d751d4da77c912f11971ea6d158/930-ng-license-manager/src/test/java/io/harness/subscription/services/CreditCardServiceImplTest.java#L58)):*
<details><summary><i>Previously, kapil-harness (Kapil Choudhary) wrote…</i></summary><blockquote>

Missing assert here
</blockquote></details>

I added an assert, but the real test is in the header. Here I am testing that exceptions are thrown.
  @Test(expected = BadRequestException.class)

**Review by kapil-harness — APPROVED** [review]
> Reviewed 4 of 9 files at r1, 7 of 10 files at r4, 3 of 3 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/48199)* status: 14 file(s) to review. 1 change requests. 2. unresolved discussions. 1 of 2 approvals obtained.

**Review by Xin-Harness — APPROVED** [review]

**Comment by thomas-grasman** [comment]
trigger review

**Pull request merged**